### PR TITLE
Request for comments - Pool tier placement.

### PR DESCRIPTION
**Pull request by kmajzero**
This is a request-for-comments patch to implement tier placement mechanism on a multi-tiered DAOS.

As multiple tiers are not implemented yet, this PR is only for review purposes.

### CODE REVIEW

File: src/include/daos_pool.h [hunk 1]
```diff
@@ -49,11 +49,12 @@ typedef struct {
 	int			foo;
 } daos_target_perf_t;
 
-enum {
+typedef enum {
 	DAOS_MEDIA_SCM	= 0,
 	DAOS_MEDIA_NVME,
+	DAOS_MEDIA_OPTANE,
```
Comment (NiuYawei) [inline src/include/daos_pool.h:55]: Maybe we can replace the 'DAOS_MEDIA_NVME' to 'DAOS_MEDIA_NVME_PERF' and 'DAOS_MEDIA_NVME_CAPACITY'?

File: src/include/daos_pool.h [hunk 2]
```diff
@@ -49,11 +49,12 @@ typedef struct {
...
 	DAOS_MEDIA_MAX
-};
+} daos_media_type_t;
```
Comment (NiuYawei) [inline src/include/daos_pool.h:57]: No new typedefs, according to our coding style.

File: src/include/daos_prop.h [hunk 3]
```diff
@@ -66,6 +66,10 @@ enum daos_pool_props {
 	 * The pool svc rank list.
 	 */
 	DAOS_PROP_PO_SVC_LIST,
+	/**
+	 * Tier selection policy
```
Comment (NiuYawei) [inline src/include/daos_prop.h:70]: Media selection policy

File: src/include/daos/policy.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * tier placement policy
+ */
+typedef enum tier_policy {
+        TIER_POLICY_NOT_EXIST = -1,
+        TIER_POLICY_DEFAULT,
+        TIER_POLICY_IO_SIZE,
+        TIER_POLICY_WRITE_INTENSIVITY,
+        TIER_POLICY_MAX
+} tier_policy_t;
```
Comment (NiuYawei) [inline src/include/daos/policy.h:25]: No new typedefs.

File: src/include/daos/policy.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#ifndef __DAOS_POLICY_H__
+#define __DAOS_POLICY_H__
+#include <daos/common.h>
+
+/**
+ * tier placement policy
+ */
+typedef enum tier_policy {
+        TIER_POLICY_NOT_EXIST = -1,
+        TIER_POLICY_DEFAULT,
```
Comment (NiuYawei) [inline src/include/daos/policy.h:21]: What's the DEFAULT policy? Isn't it one of the following policies you listed?

Just FYI, we had some planed policies in NVMe support design phase (not implemented):
- IO characteristic: Selecting media by io size;
- SCM caching: Absorbing all front IOs by SCM, migrating to NVMe by background aggregation;
- User driven: User provide hints by DAOS API to influence media selection;
- Auto tuning: analyze IO history data to predict IO workload and do proper media selection;

File: src/include/daos/policy.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifndef __DAOS_POLICY_H__
+#define __DAOS_POLICY_H__
+#include <daos/common.h>
+
+/**
+ * tier placement policy
+ */
+typedef enum tier_policy {
+        TIER_POLICY_NOT_EXIST = -1,
```
Comment (NiuYawei) [inline src/include/daos/policy.h:20]: This looks not necessary, btw, rename to DAOS_MEDIA_POLICY_XXX?

File: src/include/daos_srv/pool.h [hunk 7]
```diff
@@ -34,6 +35,7 @@ struct ds_pool {
 	struct pool_map	       *sp_map;
 	uint32_t		sp_map_version;	/* temporary */
 	uint64_t		sp_reclaim;
+	tier_policy_t		sp_policy; 	/* pool tiering policy */
```
Comment (NiuYawei) [inline src/include/daos_srv/pool.h:38]: Is an integer sufficient for describing a policy? Maybe we need a data structure to contain more parameters like what io size for SCM, what io size for performance tier, what io size for capacity tier, etc. (and the io size criteria could be different for client IO and aggregation IO)

If we expand it to a data structure, we may need to duplicate it in each ds_pool_child. (in this shared nothing architecture, 'ds_pool' is maintained & accessed by xstream 0, 'ds_pool_child' is for each VOS xstream.

File: src/include/daos_srv/vos.h [hunk 8]
```diff
@@ -1051,14 +1051,23 @@ enum vos_pool_opc {
...
+	/** Set pool tiering policy, needs a vos_ctl_set_policy_param struct
+	 *  object passed */
+	VOS_PO_CTL_SET_POLICY,
 };
 
+typedef struct {
+	tier_policy_t policy_index;
+	unsigned int  policy_io_size_low;
+	unsigned int  policy_io_size_high;
+} vos_ctl_set_policy_param;
```
Comment (NiuYawei) [inline src/include/daos_srv/vos.h:1063]: This could be the 'policy descriptor' I mentioned in prior comment?

File: src/mgmt/srv_drpc.c [hunk 9]
```diff
@@ -259,6 +260,11 @@ create_pool_props(daos_prop_t **out_prop, char *owner, char *owner_grp,
 		idx++;
 	}
 
+	/* pool tiering policy */
+	new_prop->dpp_entries[idx].dpe_type = DAOS_PROP_PO_POLICY;
+	new_prop->dpp_entries[idx].dpe_val = policy;
```
Comment (NiuYawei) [inline src/mgmt/srv_drpc.c:265]: Just keep in mind that we may need to support more configurable values for each policy, like io sizes for SCM or performance tier or capacity tier, etc? (It's not necessary being implemented in this patch)

File: src/pool/srv_internal.h [hunk 10]
```diff
@@ -46,6 +46,7 @@ struct pool_iv_prop {
 	char		pip_label[DAOS_PROP_LABEL_MAX_LEN];
 	char		pip_owner[DAOS_ACL_MAX_PRINCIPAL_BUF_LEN];
 	char		pip_owner_grp[DAOS_ACL_MAX_PRINCIPAL_BUF_LEN];
+	uint32_t	pip_policy;
```
Comment (NiuYawei) [inline src/pool/srv_internal.h:49]: The prop entry value is in uint64_t, we'd keep it consistent here.

File: src/pool/srv_target.c [hunk 11]
```diff
@@ -1281,11 +1282,39 @@ ds_pool_tgt_query_aggregator(crt_rpc_t *source, crt_rpc_t *result, void *priv)
...
+update_vos_prop_on_targets(void *in)
+{
+	struct ds_pool 			*pool = (struct ds_pool *)in;
+	struct ds_pool_child 		*child = NULL;
+	vos_ctl_set_policy_param 	param;
+	int 				ret = 0;
+
+	child = ds_pool_child_lookup(pool->sp_uuid);
+	if (child == NULL)
+		return 0; //no child created yet?
```
Comment (NiuYawei) [inline src/pool/srv_target.c:1295]: Return an error.

File: src/rdb/rdb.c [hunk 12]
```diff
@@ -373,6 +375,7 @@ rdb_start(const char *path, const uuid_t uuid, struct rdb_cbs *cbs, void *arg,
 			path, DP_RC(rc));
 		goto err;
 	}
+
 	ABT_thread_yield();
 
 	rc = vos_cont_open(pool, (unsigned char *)uuid, &mc);
```
Comment (NiuYawei) [inline src/rdb/rdb.c:381]: Avoid unnecessary changes in this file.

File: src/vos/vos_policy.c [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <daos/policy.h>
+#include "vos_policy.h"
+
+
+/* policy functions definitions */
+
+/* default policy - former "vos_media_select" function */
+static daos_media_type_t
+policy_default(struct vos_pool *pool, daos_iod_type_t type, daos_size_t size)
```
Comment (NiuYawei) [inline src/vos/vos_policy.c:21]: ah, I see, this is named 'default' policy. I think current policy should be merged into io size policy, right?

File: src/vos/vos_policy.c [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                                                        daos_iod_type_t,
+                                                        daos_size_t) =
+                   {policy_default, policy_io_size, policy_write_intensivity};
+
+
+
+
+daos_media_type_t
+vos_policy_media_select(struct vos_pool *pool, daos_iod_type_t type,
+                        daos_size_t size)
```
Comment (NiuYawei) [inline src/vos/vos_policy.c:68]: It needs one more input: "enum vos_io_stream ios".

**Comment by NiuYawei** [comment]
@tanabarr , @mjmac Could you take a look at the CP changes to see if they are on the right direction? Thanks in advance.

File: src/control/cmd/dmg/pool.go [hunk 15]
```diff
@@ -90,6 +108,11 @@ func (cmd *PoolCreateCmd) Execute(args []string) error {
 		return errors.Wrap(err, "parsing rank list")
 	}
 
+	req.Policy, err = ParsePolicy(cmd.Policy)
```
Comment (tanabarr) [inline src/control/cmd/dmg/pool.go:111]: move to control API as recommended above

File: src/control/drpc/property.go [hunk 16]
```diff
@@ -45,3 +48,12 @@ const (
 	// PoolSelfHealingAutoRebuild sets the self-healing strategy to auto-rebuild.
 	PoolSelfHealingAutoRebuild = C.DAOS_SELF_HEAL_AUTO_REBUILD
 )
+
+const (
```
Comment (tanabarr) [inline src/control/drpc/property.go:52]: convert to named const e.g. PoolPolicy PoolPolicyDefault = C.TIER_POLICY_DEFAULT then ParsePolicy(string) (PoolPolicy, error) can map string to typed const if found

File: src/control/lib/control/pool.go [hunk 17]
```diff
@@ -134,6 +136,7 @@ type (
 		Ranks     []system.Rank
 		ScmBytes  uint64
 		NvmeBytes uint64
+		Policy    uint32
```
Comment (tanabarr) [inline src/control/lib/control/pool.go:139]: type PoolPolicy

**Review by tanabarr — COMMENTED** [review]
nothing blocking, just suggestions

File: src/proto/mgmt/pool.proto [hunk 18]
```diff
@@ -33,6 +33,7 @@ message PoolCreateReq {
 	repeated uint32 ranks = 12; // target ranks (manual config)
 	uint64 scmbytes = 13; // SCM size in bytes (manual config)
 	uint64 nvmebytes = 14; // NVMe size in bytes (manual config)
+	uint32 policy = 15; // tier placement policy (manual config)
```
Comment (mjmac) [inline src/proto/mgmt/pool.proto:36]: Might be better to name this tier_policy so that we don't need to rename it later if we add support for other kinds of policies. We're trying to standardize the protobuf field names on snake_case as it results in more consistent JSON serialization.

**Review by mjmac — COMMENTED** [review]
Couple of minor comments... Overall, looks good but I do wonder if we should name this `tier_policy` rather than just `policy` in order to avoid conflict with future work that might also implement some kind of policy-based feature.

**Comment by tanabarr** [comment]
@kmajzero it would be nice if we could try to avoid force pushing, if you push the new commits without force pushing then reviewers can see the new changes rather than having to rereview the whole PR. Thanks in advance.

File: src/control/lib/control/pool.go [hunk 19]
```diff
@@ -708,3 +711,23 @@ func PoolReintegrate(ctx context.Context, rpcClient UnaryInvoker, req *PoolReint
...
+// ParsePolicy will parse the incoming policy name string and return
+// a policy index used in the DAOS engine.
+// Returns an error if string cannot be found in the map
+func ParsePolicy(stringPolicy string) (policy uint32, err error) {
+
+	var policyMap = map[string]drpc.PoolPolicy{
+		"default":           drpc.PoolPolicyDefault,
+		"io_size":           drpc.PoolPolicyIoSize,
+		"write_intensivity": drpc.PoolPolicyWriteIntensivity,
+	}
```
Comment (tanabarr) [inline src/control/lib/control/pool.go:725]: you can collapse this into policy, found := map[string]drpc.PoolPolicy{...}[stringPolicy]

File: src/control/go.mod [hunk 20]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/armon/go-metrics v0.3.6 // indirect
 	github.com/dustin/go-humanize v1.0.0
 	github.com/fatih/color v1.10.0 // indirect
+	github.com/golang/protobuf v1.5.2
```
Comment (tanabarr) [inline src/control/go.mod:10]: I think @mjmac has moved all the imports to golang.org repo, he can confirm .

**Review by tanabarr — COMMENTED** [review]
changes on the go side look good to me, just need to update the protobuf package imports in line with @mjmac 's recent commit on master

File: src/include/daos/policy.h [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+ * include/daos/policy.h
+ */
+
+#ifndef __DAOS_POLICY_H__
+#define __DAOS_POLICY_H__
+#include <daos/common.h>
+
+/** tier placement policy */
+enum tier_policy_t {
+        DAOS_MEDIA_POLICY_DEFAULT,
```
Comment (NiuYawei) [inline src/include/daos/policy.h:18]: I think default policy is "io size" policy, right?

File: src/vos/vos_policy.c (hunk 13 above)
Comment (NiuYawei) [reply src/vos/vos_policy.c:21]: Please get rid of this 'default' policy, and use 'io size' policy as default policy.

**Comment by johannlombardi** [comment]
Please move to a branch in the DAOS repo and resubmit a new PR so that testing is triggered.
